src/ring_theory/maps.lean

```diff
@@ -69,130 +43,19 @@ lemma map_zero_iff {x : R} : Hs x = 0 ↔ x = 0 :=
 
 end ring_equiv
 
-/-- A ring antiisomorphism -/
-structure ring_anti_equiv [ring R] [ring F] extends R ≃ F :=
-[anti_hom : is_ring_anti_hom to_fun]
-
-namespace ring_anti_equiv
-
-variables {R F} [ring R] [ring F] (Hs : ring_anti_equiv R F) (x y : R)
-
-instance : has_coe_to_fun (ring_anti_equiv R F) :=
-⟨_, λ Hs, Hs.to_fun⟩
-
-instance : is_ring_anti_hom Hs := Hs.anti_hom
-
-lemma map_add : Hs (x + y) = Hs x + Hs y :=
-is_ring_anti_hom.map_add Hs
-
-lemma map_zero : Hs 0 = 0 :=
-is_ring_anti_hom.map_zero Hs
-
-lemma map_neg : Hs (-x) = -Hs x :=
-is_ring_anti_hom.map_neg Hs
-
-lemma map_sub : Hs (x - y) = Hs x - Hs y :=
-is_ring_anti_hom.map_sub Hs
-
-lemma map_mul : Hs (x * y) = Hs y * Hs x :=
-is_ring_anti_hom.map_mul Hs
-
-lemma map_one : Hs 1 = 1 :=
-is_ring_anti_hom.map_one Hs
-
-lemma map_neg_one : Hs (-1) = -1 :=
-Hs.map_one ▸ Hs.map_neg 1
-
-lemma bijective : function.bijective Hs := Hs.to_equiv.bijective
-
-lemma map_zero_iff {x : R} : Hs x = 0 ↔ x = 0 :=
-⟨λ H, Hs.bijective.1 $ H.symm ▸ Hs.map_zero.symm,
-λ H, H.symm ▸ Hs.map_zero⟩
-
-end ring_anti_equiv
-
 /-- A ring involution -/
-structure ring_invo [ring R] :=
-(to_fun : R → R)
-[anti_hom : is_ring_anti_hom to_fun]
-(to_fun_to_fun : ∀ x, to_fun (to_fun x) = x)
+structure ring_invo [ring R] extends R ≃+* opposite R :=
+(to_fun_to_fun : ∀ x, (to_fun (to_fun x).unop).unop = x)
```

Comment:
This is not the minimal definition (it includes `inv_fun`, `left_inv`, and `right_inv`). The minimal def is
```
structure ring_invo [semiring R] :=
(to_fun : R → opposite R)
(map_mul' : ∀ x y, to_fun (x * y) = to_fun x * to_fun y)
(map_add' : ∀ x y, to_fun (x + y) = to_fun x + to_fun y)
(involution' : ∀ x, (to_fun (to_fun x).unop).unop = x)
```
BTW, is it more useful as a map `R → opposite R` or `R → R` with `f.to_equiv : R →+* opposite R`?

Reply:
After trying both ways, it seems nicer to leave `ring_invo` as extending `R ≃+* opposite R`, but to provide a cheaper constructor separately. I've added some additional API to make `ring_invo` easier to use.